lib/rust/frp/src/nodes.rs

```diff
@@ -1583,6 +1594,49 @@ impl<T: EventOutput> stream::EventConsumer<Output<T>> for OwnedTrace<T> {
 
 
 
+// ===============
+// === Profile ===
+// ===============
+
+#[derive(Clone, Debug)]
+pub struct ProfileData<T> {
+    #[allow(dead_code)]
+    /// This is not accessed in this implementation but it needs to be kept so the source struct
+    /// stays alive at least as long as this struct.
+    src: T,
+}
+pub type OwnedProfile<T> = stream::Node<ProfileData<T>>;
+pub type Profile<T> = stream::WeakNode<ProfileData<T>>;
+
+impl<T: EventOutput> HasOutput for ProfileData<T> {
+    type Output = Output<T>;
+}
+
+impl<T: EventOutput> OwnedProfile<T> {
+    /// Constructor.
+    pub fn new(label: Label, src1: &T) -> Self {
+        let src = src1.clone_ref();
+        let def = ProfileData { src };
+        Self::construct_and_connect(label, src1, def)
+    }
+}
+
+impl<T: EventOutput> stream::EventConsumer<Output<T>> for OwnedProfile<T> {
+    fn on_event(&self, stack: CallStack, event: &Output<T>) {
+        use profiler::internal::Profiler;
+
+        let label = profiler::internal::Label(self.label());
```

Comment:
Since this is a special case that relies on the internal `profiler` APIs, we need to take care it doesn't violate the usual `profiler` invariants; it should have a test that profiles some FRP events and parses the data into `profiler_data` Measurements.

i don't see where this label ultimately comes from; is it possible to pack origin-in-code info into the str to fit the "label (file:line)" format? Without that info it will panic `profiler_data` parsing. If getting `file:line` info into the labels isn't feasible, we can be less strict in `profiler_data::parse` and accept labels that fail to parse as literal labels with no `CodePos` info.

Reply:
`we need to take care it doesn't violate the usual profiler invariants` - could you elaborate more bout what are the "usual profiler invariants", please? Are you talking about the label (file:line) format, or about anything else too?

Reply:
I think this is about the label format. I would try to relax the parsing requirements. It would be quite tricky to get the correct `file:line` information into the profiling node (in a performant way) I think.

Reply:
Because performance is a major design concern for the event log, it is "denormalized"; **it is possible for the event log to represent a lot of things that are invalid** (and can't be represented by `Measurement`). Many of [the possible errors](https://github.com/enso-org/enso/blob/a3914f33c13b94cd7bfba69b8c818db39327d381/lib/rust/profiler/data/src/parse.rs#L97) relate to profiler state transitions, and relationships between profilers. For example, a profiler must not pause when it is already in the paused state. As another example, the event log could contain: start A, start B, end A (without pausing or ending B first); that is not a valid sequence of events, and can't be parsed into Measurements by `profiler_data`.

It is impossible to produce most cases of invalid data with the `#[profile]` API (the only exceptions have to do with naively instrumenting low-level async executor facilities), and difficult to do with the public low-level API (you could for example explicitly `drop` `Started` guards out of order); however **it would be easy to create invalid data with the `internal` API**, which allows almost anything the event log can represent. So I think on the rare occasion it is necessary, **usage of the `profiler::internal` APIs should be accompanied by `profiler_data` parse tests** to check that the usage is valid.

This case is fairly simple; the only issue that a `profiler_data` test would catch right now is the `Label` format. But correct usage now could be broken by small changes here, to the `profiler` implementation, or even to the data model `profiler::data` enforces.

Reply:
@kazcw this is an amazing explanation. Would you be so nice and add it as docs to the profiler framework somewhere? It gives so many insights. Ofc if it is so clearly written in docs already, skip my request. I haven't seen it there though, but my brain easily forgets things :)